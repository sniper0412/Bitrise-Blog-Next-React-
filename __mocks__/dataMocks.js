export const mockAuthors = [
  {
    slug: 'author-1',
    firstName: 'Bender',
    lastName: 'Rodr√≠guez',
    profileImage: 'http://bender.img',
    title: 'Chief Bending Robot',
    bio: 'A fictional character who is one of the main characters in the animated television series Futurama'
  }
];

export const mockCategories = [
  {
    slug: 'category-1',
    name: 'category 1'
  },
  {
    slug: 'category-2',
    name: 'category 2'
  }
];

export const mockTags = [
  {
    slug: 'tag-1',
    name: 'tag 1'
  },
  {
    slug: 'tag-2',
    name: 'tag 2'
  }
];

export const mockPosts = [
  {
    author: mockAuthors[0],
    slug: 'posts-1',
    seoTitle: 'post 1',
    metaDescription: 'Post 1 description',
    featuredImage: 'http://post-1.img',
    title: 'Post 1',
    body: '<h1>Post 1 body</h1>',
    categories: [mockCategories[0], mockCategories[1]],
    tags: [mockTags[0], mockTags[1]],
    published: 'Wed Jan 02 2019 16:50:20 GMT+0100',
    summary: 'Post 1 summary'
  }
];
