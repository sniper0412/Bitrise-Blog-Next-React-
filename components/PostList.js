import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import mapKeys from 'lodash/mapKeys';
import snakeCase from 'lodash/snakeCase';
import { format } from 'url';

import { PostType } from '../types';

import PostSummary from './PostSummary';
import { camlizeKeysDeep } from '../utils';

export default class PostList extends React.Component {
  static propTypes = {
    initialPosts: PropTypes.arrayOf(PropTypes.shape(PostType)),
    count: PropTypes.number.isRequired,
    filters: PropTypes.shape({
      tagSlug: PropTypes.string,
      categorySlug: PropTypes.string,
      authorSlug: PropTypes.string,
      query: PropTypes.string
    })
  };

  state = {
    page: 2, // Butter page count starts at 1, the first page is loaded by the server
    isLoading: false,
    posts: []
  };

  async loadMorePosts() {
    const { filters } = this.props;
    const { page, isLoading } = this.state;
    if (isLoading) return;

    this.setState({ isLoading: true });

    const url = format({
      pathname: '/fetch-posts',
      query: { ...mapKeys(filters, (_, key) => snakeCase(key)), page }
    });

    const posts = (await fetch(url).then(res => res.json())).map(camlizeKeysDeep);

    this.setState({ isLoading: false, page: page + 1, posts: this.state.posts.concat(posts) });
  }

  render() {
    const { count, initialPosts } = this.props;
    const { posts } = this.state;

    const allPosts = [...initialPosts, ...posts];

    return (
      <Fragment>
        <div className="articles">
          {allPosts.map((post, key) => (
            <PostSummary key={key} {...post} defaultImagePath="/static/img/post-default-img.jpg" />
          ))}
        </div>

        {count > allPosts.length && (
          <div className="default-button load-more" onClick={() => this.loadMorePosts()}>
            <span className="default-button__content">Load more</span>
          </div>
        )}
      </Fragment>
    );
  }
}
