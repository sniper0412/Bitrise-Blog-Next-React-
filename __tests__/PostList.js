import renderer from 'react-test-renderer';

import { mockPosts } from 'dataMocks';
import PostList from '../components/PostList';

describe('PostList component', () => {
  it('renders correctly when it can load more', () => {
    const posts = [mockPosts[0]];
    const tree = renderer.create(<PostList initialPosts={posts} count={100} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders correctly when it cannot load more', () => {
    const posts = [mockPosts[0]];
    const tree = renderer.create(<PostList initialPosts={posts} count={1} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
