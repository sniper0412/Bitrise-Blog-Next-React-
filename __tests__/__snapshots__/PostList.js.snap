// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostList component renders correctly when it can load more 1`] = `
Array [
  <div
    className="articles"
  >
    <div
      className="article-container"
    >
      <a
        href="/posts/posts-1"
        title="Post 1"
      >
        <div
          className="header-container"
        >
          <p
            className="category"
          >
            category 1
          </p>
          <p
            className="date"
          >
            January 2, 2019
          </p>
        </div>
        <div
          className="image-container"
          style={
            Object {
              "backgroundImage": "url(http://post-1.img)",
            }
          }
        />
        <div
          className="article"
        >
          <h2>
            Post 1
          </h2>
          <p>
            Post 1 summary
          </p>
        </div>
      </a>
    </div>
  </div>,
  <div
    className="default-button load-more"
    onClick={[Function]}
  >
    <span
      className="default-button__content"
    >
      Load more
    </span>
  </div>,
]
`;

exports[`PostList component renders correctly when it cannot load more 1`] = `
<div
  className="articles"
>
  <div
    className="article-container"
  >
    <a
      href="/posts/posts-1"
      title="Post 1"
    >
      <div
        className="header-container"
      >
        <p
          className="category"
        >
          category 1
        </p>
        <p
          className="date"
        >
          January 2, 2019
        </p>
      </div>
      <div
        className="image-container"
        style={
          Object {
            "backgroundImage": "url(http://post-1.img)",
          }
        }
      />
      <div
        className="article"
      >
        <h2>
          Post 1
        </h2>
        <p>
          Post 1 summary
        </p>
      </div>
    </a>
  </div>
</div>
`;
